### PR TITLE
BaSP-TG: Task Endpoints (GetOne, Update, Delete)

### PR DESCRIPTION
**Pull request by Martinzancocchia**
**

> endpoints added to tasks:

**

- getOne
- Update
- Delete

The validations are there, only that when I want to use them I can't start the server, could you see what I'm missing and leave me a comment?

### CODE REVIEW

File: src/controllers/task.js [hunk 1]
```diff
@@ -0,0 +1,143 @@
+import Models from '../models/Task';
+// import task from '../validations/task';
```
Comment (ignacioaristo) [inline src/controllers/task.js:2]: Remove this comment.

File: src/controllers/task.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+export const getAll = async (req, res) => {
+  try {
+    const tasks = await Models.find();
+    return res.status(200).json({
+      message: 'Tasks found',
+      data: tasks,
+      error: false,
```
Comment (ignacioaristo) [inline src/controllers/task.js:10]: I think this error is not necessary.

File: src/controllers/task.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    });
+  } catch (err) {
+    return res.json({
+      message: 'an error ocurred',
```
Comment (ignacioaristo) [inline src/controllers/task.js:14]: We can handle the error in a better way, for example, `message: 'an error ocurred: ${err}',`

File: src/controllers/task.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      error: err,
```
Comment (ignacioaristo) [inline src/controllers/task.js:15]: With the previous commet, we don't need this.

File: src/controllers/task.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    });
+  }
+};
+
+export const getTaskById = async (req, res) => {
+  try {
+    const { id } = req.params;
+    if (!id.match(/^[0-9a-fA-F]{24}$/)) {
```
Comment (ignacioaristo) [inline src/controllers/task.js:23]: For what is this for?

File: src/controllers/task.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      return res.status(400).json({
+        msg: 'missing or wrong id parameter',
+        data: undefined,
```
Comment (ignacioaristo) [inline src/controllers/task.js:26]: This is not necessary.

File: src/controllers/task.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        error: true,
+      });
+    }
+    const task = await Models.findById(id);
+    if (!task) {
+      return res.status(404).json({
+        msg: 'The task has not been found',
+        data: undefined,
```
Comment (ignacioaristo) [inline src/controllers/task.js:34]: Same here.

File: src/controllers/task.js [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+  } catch (error) {
+    return res.json({
+      msg: 'There has been an error:',
+      data: error,
+      error: true,
+    });
+  }
+};
+
+// export const createTask = async (req, res) => {
```
Comment (ignacioaristo) [inline src/controllers/task.js:52]: If we don't need this, remove all the comments lines.

File: src/controllers/task.js [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      });
+    }
+    return res.status(200).json({
+      msg: 'The task has been found',
+      data: task,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
+      msg: 'There has been an error:',
```
Comment (ignacioaristo) [inline src/controllers/task.js:45]: Let's handle the error message like the previous comment.

File: src/controllers/task.js [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+      data: error,
+      error: true,
+    });
+  }
+};
+
+// export {
```
Comment (ignacioaristo) [inline src/controllers/task.js:139]: I prefer use the export. But if you don't delete these commets.

File: src/controllers/task.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+      });
+    }
+    return res.status(200).json({
+      msg: 'The task has been Updated',
+      data: undefined,
+      error: false,
+    });
+  } catch (error) {
+    return res.json({
+      msg: 'There has been an error',
```
Comment (ignacioaristo) [inline src/controllers/task.js:100]: Same error handling and delete the following props.

File: src/index.js [hunk 12]
```diff
@@ -1,25 +1,27 @@
...
+import routes from './routes/index';
 
 // use "require" to import JSON files
-const tasksRouter = require('./resources/tasks');
-const employees = require('./resources/employees');
-const timeSheetsRouter = require('./resources/time-sheets');
-const superAdmin = require('./resources/super-admins');
-const projectsRouter = require('./resources/projects');
-const adminRouter = require('./resources/admins');
+// const tasksRouter = require('./resources/tasks');
```
Comment (ignacioaristo) [inline src/index.js:7]: Why comment these lines? The server will work if you uncomment this.

File: src/index.js [hunk 13]
```diff
@@ -1,25 +1,27 @@
...
 
 app.use(express.json());
-app.use('/projects', projectsRouter);
-app.use('/tasks', tasksRouter);
-app.use('/timeSheets', timeSheetsRouter);
-app.use('/employees', employees);
-app.use('/superAdmin', superAdmin);
-app.use('/admins', adminRouter);
+app.use('/', routes);
+// app.use('/projects', projectsRouter);
```
Comment (ignacioaristo) [inline src/index.js:19]: The same here.

File: src/routes/task.js [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import express from 'express';
+// import validations from '../validations/task';
```
Comment (ignacioaristo) [inline src/routes/task.js:2]: Delete all the comment lines

File: src/routes/task.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+// import { updateTaskValidation } from '../validations/task';
+
+const router = express.Router();
+
+router
+  .delete('/:id', deleteTask)
+//   .post('/', createTask)
+  .get('/:id', getTaskById)
+  .put('/:id', editTask)
+  .get('/', getAll);
```
Comment (ignacioaristo) [inline src/routes/task.js:15]: This don't belong to this ticket.

**Review by ignacioaristo — CHANGES_REQUESTED** [review]

File: src/controllers/task.js (hunk 1 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:2]: done

File: src/controllers/task.js (hunk 4 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:15]: done

File: src/controllers/task.js (hunk 5 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:23]: fixed

File: src/controllers/task.js (hunk 6 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:26]: fixed

File: src/controllers/task.js (hunk 7 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:34]: done

File: src/controllers/task.js (hunk 2 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:10]: done

File: src/controllers/task.js (hunk 3 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:14]: done

File: src/controllers/task.js (hunk 4 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:15]: done

File: src/controllers/task.js (hunk 11 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:100]: done

File: src/index.js (hunk 12 above)
Comment (Martinzancocchia) [reply src/index.js:7]: done

File: src/index.js (hunk 13 above)
Comment (Martinzancocchia) [reply src/index.js:19]: done

File: src/routes/task.js (hunk 14 above)
Comment (Martinzancocchia) [reply src/routes/task.js:2]: fixed

File: src/routes/task.js (hunk 15 above)
Comment (Martinzancocchia) [reply src/routes/task.js:15]: fixed!

File: src/controllers/task.js (hunk 8 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:52]: done

File: src/controllers/task.js (hunk 9 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:45]: done

File: src/controllers/task.js (hunk 10 above)
Comment (Martinzancocchia) [reply src/controllers/task.js:139]: fixed

File: src/index.js (hunk 12 above)
Comment (ignacioaristo) [reply src/index.js:7]: i agree, but we can't have comment code. So for now you have to leave it, the other devs have to remove it from their PRs.

File: src/routes/task.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import express from 'express';
+// import validations from '../validations/task';
+import {
+  deleteTask, updateTask, getOneTask,
+} from '../controllers/task';
+// import { updateTaskValidation } fr|om '../validations/task';
```
Comment (ignacioaristo) [inline src/routes/task.js:6]: Delete this comment line

**Review by ignacioaristo — CHANGES_REQUESTED** [review]

File: src/validations/task.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
+import Joi from 'joi';
+
+const taskValidateUpdate = Joi.object({
+  description: Joi.string().valid(
```
Comment (paurinaldi) [inline src/validations/task.js:4]: Since this is the only field, it should be required in order to update.

**Review by paurinaldi — CHANGES_REQUESTED** [review]

File: src/models/Task.js [hunk 18]
```diff
@@ -0,0 +1,13 @@
+import mongoose from 'mongoose';
+
+const { Schema } = mongoose;
+
+const taskSchema = new Schema({
+  description: {
+    type: String,
+    required: true,
+    lowercase: true,
```
Comment (paurinaldi) [inline src/models/Task.js:9]: Why should we add this line?
I suggest you talk to sabri and you both establish the same shema for tasks.

File: src/validations/task.js (hunk 17 above)
Comment (Martinzancocchia) [reply src/validations/task.js:4]: Sorry! I just got out of an exam, when I get home I'll change it!

Comment (Martinzancocchia) [reply src/validations/task.js:4]: Sorry! I just got out of an exam, when I get home I'll change it!

File: src/routes/task.js (hunk 16 above)
Comment (Martinzancocchia) [reply src/routes/task.js]: done

File: src/routes/index.js [hunk 19]
```diff
@@ -1,8 +1,10 @@
 import express from 'express';
+import taskRouter from './task';
 import admin from './admins';
 
 const router = express.Router();
```
Comment (paurinaldi) [inline src/routes/index.js:6]: please make it /tasks to keep consistency

File: src/index.js [hunk 20]
```diff
@@ -14,8 +12,8 @@ const app = express();
 const port = process.env.PORT || 3000;
 
 app.use(express.json());
+app.use('/', routes);
```
Comment (paurinaldi) [inline src/index.js:15]: we established route should be '/api'

File: src/controllers/task.js [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+export const deleteTask = async (req, res) => {
+  try {
+    const { id } = req.params;
+    const result = await Models.findByIdAndDelete(id);
+    if (!result) {
+      return res.status(404).json({
+        msg: 'The task has not been found',
+      });
+    }
+    return res.status(200).json({
```
Comment (paurinaldi) [inline src/controllers/task.js:54]: when we delete something from the database, status code should be 204

**Review by paurinaldi — CHANGES_REQUESTED** [review]

File: src/routes/index.js (hunk 19 above)
Comment (paurinaldi) [reply src/routes/index.js:9]: I still see task as the route, should be /tasks

**Review by paurinaldi — APPROVED** [review]

**Pull request merged**